src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs

```diff
@@ -1087,27 +1104,37 @@ private ushort ReadUint16(BufferedReadStream stream)
         /// Post processes the pixels into the destination image.
         /// </summary>
         /// <typeparam name="TPixel">The pixel format.</typeparam>
-        /// <returns>The <see cref="Image{TPixel}"/>.</returns>
-        private Image<TPixel> PostProcessIntoImage<TPixel>(CancellationToken cancellationToken)
+        /// <param name="image">The destination image.</param>
+        /// <param name="cancellationToken">The cancellation token for the operation.</param>
+        private void PostProcessIntoImage<TPixel>(ref Image<TPixel> image, CancellationToken cancellationToken)
             where TPixel : unmanaged, IPixel<TPixel>
         {
             if (this.ImageWidth == 0 || this.ImageHeight == 0)
             {
                 JpegThrowHelper.ThrowInvalidImageDimensions(this.ImageWidth, this.ImageHeight);
             }
 
-            var image = Image.CreateUninitialized<TPixel>(
-                this.Configuration,
-                this.ImageWidth,
-                this.ImageHeight,
-                this.Metadata);
-
-            using (var postProcessor = new JpegImagePostProcessor(this.Configuration, this))
+            if (image is null)
             {
-                postProcessor.PostProcess(image.Frames.RootFrame, cancellationToken);
+                image = Image.CreateUninitialized<TPixel>(
+                    this.Configuration,
+                    this.ImageWidth,
+                    this.ImageHeight,
+                    this.Metadata);
             }
+            else
+            {
+                if (this.ImageWidth != image.Width || this.ImageHeight != image.Height)
+                {
+                    ThrowHelper.ThrowArgumentException("The input image has an invalid size", nameof(image));
```

Comment:
```suggestion
                    ThrowHelper.ThrowArgumentException("The input image should match the dimensions of the image in the stream.", nameof(image));
```

Same for the other decoders. We can throw on this much earlier in the decoding workflow.